Updated guide on how to create custom plugins (#8055)
handsontable/handsontable: docs/next/guides/building-and-testing/plugins.md

---
title: Plugins
metaTitle: Plugins - Guide - Handsontable Documentation
permalink: /next/plugins
canonicalUrl: /plugins
tags:
  - custom plugin
  - skeleton
  - extend
---

# Plugins

## Overview

Plugin is a one great way to extend the capabilities of Handsontable. In fact, most of features available in this library are provided by plugins.

This guide will lead you, step-by-step, through the process of creating a custom extension.

### 1. Prerequisites

You need to import the following things:
- `BasePlugin` - contains the basic interface to work within Handsontable's lifecycle,
- `registerPlugin` - utility to register the plugin in Handsontable plugins registry.

Both provide the basic interface necessary to create a reusable plugin.

```js
import { BasePlugin, registerPlugin } from 'handsontable/plugins';
```

### 2. Extend BasePlugin
Extending our `BasePlugin` is a recommended way to start creating your next extension.
In this built-in interface, we care about many factors such as backward compatibility, memory leaks preventions and proper binding plugin's instance with Handsontable.

```js
export class CustomPlugin extends BasePlugin {
  /**

>    * Define an unique plugin's key.

Reviewer: an unique -> a unique

>    * Handsontable will use that string to register the plugin under that alias.

Reviewer: will use -> uses

   * Moreover, you might also use the same parameter to recognize the plugin's

>    * options passed through the setting object at Handsontable initialization.

Reviewer: setting object -> Settings object

   *
   * @returns {string}
   */
  static get PLUGIN_KEY() {
    return 'customPlugin';
  }

  /**
   * Extend the default constructor and define internal properties for your plugin.
   *
   * @param {Handsontable} hotInstance
   */
  constructor(hotInstance) {

>     super(hotInstance);

Reviewer 2: Through this tutorial we use `this.hot` few times. But we never explain where it comes from. Maybe we should section or an explanation that BasePlugin provides few properties for us to use?

There are index mappers as well, we may link those to the index mapper description to keep this page short

Author: Done! However, I cannot link to the index-mapper part in Core (API ref) because the code snippet does not convert shortcodes `[...](...)` to links.

    // Initialize all your public properties in the class' constructor.
    this.configuration = null;
  }

  /**
   * Checks if the plugin is enabled in the settings.
   */
  isEnabled() {

>     return !!this.hot.getSettings()[CustomPlugin.PLUGIN_KEY];

Reviewer 2: For the tutorial we should make it obvious that it is enabled if `true` or has settings. May add the explanation to the comment above
```suggestion
    const pluginSettings = this.hot.getSettings()[CustomPlugin.PLUGIN_KEY];
    return pluginSettings === true || Object.isObject(pluginSettings);
```

Author: I got the point, but I see the two problems: 
1) `Object.isObject` ❤️😄 great idea but not exists in ES
2) We do not export modularized `handsontable/helpers` where we have our own helper of `isObject` 🙄

Reviewer 2: 1. Hold my 🍺 we have to fix W3C spec first 😄 
2. Maybe lodash `_.isObject` then?
3. We may also add plugin method `validateConfig` and place some more detailed asserts there

Author: Yes, I thought about the third option too. Now there will be the "getUnifiedConfig" method.

    // Add all your plugin hooks here. It's a good idea to use arrow functions to keep the plugin as a context.
    this.addHook('afterChange', (changes, source) => this.onAfterChange(changes, source));

    // The super method assigns the this.enabled property to true.
    // It is a necessary step to update the plugin's settings properly.
    super.enablePlugin();
  }

  /**
   * The disablePlugin method is used to disable the plugin.
   */
  disablePlugin() {
    // Reset all of your plugin class properties to their default values here.
    this.configuration = null;

    // The super method takes care of clearing the hook connections and assigning the 'false' value to the 'this.enabled' property.
    super.disablePlugin();
  }

  /**
   * The updatePlugin method is called on the `afterUpdateSettings` hook
   * (unless the updateSettings method turned the plugin off).
   *
   * It should contain all the stuff your plugin needs to do to work correctly
   * after updating Handsontable instance settings.
   */
  updatePlugin() {
    // The `updatePlugin` method needs to contain all the code needed to properly re-enable the plugin.

>     // In most cases simply disabling and enabling the plugin should do the trick.

Reviewer 2: 🙈

Author: Yes, now it works ;)

PS. If you write about UndoRedo - that's the other problem :trollface:

>     this.disablePlugin();
>     this.enablePlugin();

Reviewer 2: This is not the best practice. Let's teach our users to do it right.

```suggestion
    const pluginSettings = this.hot.getSettings()[CustomPlugin.PLUGIN_KEY];
    if (pluginSettings === false && !this.enabled) {
      this.disablePlugin();
    }
    
    if (pluginSettings === true || !this.enabled) {
      this.enablePlugin();
    }
    
    if (Object.isObject(pluginSettings)) {
      this.configuration = pluginSettings;
      // update some internal state if needed
    }
```

Author: Thank you for the idea. I cannot merge it because the suggested code snippet is not compatible with the changes I put on the example anymore.

    super.updatePlugin();
  }

  /**
   * Define your external methods.
   */
  externalMethodExample() {
    // Method definition.
  }

  /**
   * The afterChange hook callback.
   *
   * @param {CellChange[]} changes An array of changes.
   * @param {string} source Describes the source of the change.
   */
  onAfterChange(changes, source) {
    // afterChange callback goes here.
  }

  /**
   * The `destroy` method is the best place to clean up any instances,
   * objects or index mappers created during the plugin's lifecycle.
   */
  destroy() {

>     // The super method takes care of de-assigning the event callbacks,

Reviewer: The super method -> This method

Author: The original text is referring to the nature of class inheritance.
Running `super.xyz()` in the child class will invoke the `xyz` method from the parent class (simplified description).
Although, sure, we can mention that's the `BasePlugin.disablePlugin` method.

    // plugin hooks and clearing all the plugin properties.
    super.destroy();
  }
}
```

### 3. Register CustomPlugin
Last but not least, you need to register the plugin. This is the only way to use plugins in Handsontable.

> There are two options.

Reviewer: There are two options. -> There are two ways to do that.

> - You can define a static getter named `PLUGIN_KEY` that the `registerPlugin` utility will use as the plugin's alias.

Reviewer: will use -> uses

> - You can define a static getter named `PLUGIN_KEY` that the `registerPlugin` utility will use as the plugin's alias.

Reviewer 2: Something like "As we did in the example before" would assure the user that the getter was a good idea and now he should choose this option.

  ```js
  // You need to register your plugin in order to use it within Handsontable.
  registerPlugin(CustomPlugin);
  ```

> - You can also use the alternative alias

Reviewer 2: It may also be used to alias the plugin name? If for some reason the name used in `PLUGIN_KEY` is not desireable

Author: No, I had to read my sentence twice to catch the mistake :see_no_evil:

If the end-developer use a custom alias, that string will be used to register the plugin - even if the plugin's class defines `PLUGIN_KEY`.

  ```js
  registerPlugin('CustomAlias', CustomPlugin);
  ```

### 4. Use plugin in Handsontable

```js
import Handsontable from 'handsontable';
import { CustomPlugin } from './customPlugin';

const hotInstance = new Handsontable(container, {

>   [CustomPlugin.PLUGIN_KEY]: true

Reviewer 2: We should add that this may be an object with configuration. Currently it looks like we need `true` to enable the plugin and we don't mention how to provide additional settings to the plugin. We can do it here or create another section about plugin settings

});
```

### 5. Get a reference to the plugin's instance

> If you would like to use plugin's API you might use [`getPlugin`](../api/core/#getplugin) method to get a refernce to the plugin's instance.

Reviewer: refernce -> reference

> If you would like to use plugin's API you might use [`getPlugin`](../api/core/#getplugin) method to get a refernce to the plugin's instance.

Reviewer: To use the plugin's API you may use ...

## Overview

Plugin is a one great way to extend the capabilities of Handsontable. In fact, most of features available in this library are provided by plugins.

This guide will lead you, step-by-step, through the process of creating a custom extension.

### 1. Prerequisites

You need to import the following:
- `BasePlugin` - contains the basic interface to work within Handsontable's lifecycle,
- `registerPlugin` - utility to register the plugin in Handsontable plugins registry.

> Both provide the basic interface necessary to create a reusable plugin.

Reviewer 3: ```suggestion

```

```js
import { BasePlugin, registerPlugin } from 'handsontable/plugins';
```

### 2. Extend BasePlugin
Extending our `BasePlugin` is a recommended way to start creating your next extension.

> In this built-in interface, we care about many factors such as backward compatibility, memory leaks preventions and proper binding plugin's instance with Handsontable.

Reviewer 3: ~Both provide the basic interface necessary to create a reusable plugin.~

```js
import { BasePlugin, registerPlugin } from 'handsontable/plugins';
```

\### 2. Extend the \`BasePlugin\`
The best way to start creating your own plugin is to extend the \`BasePlugin\`.

The \`BasePlugin\` interface takes care of:
\* Backward compatibility
\* Memory leak prevention
\* Properly binding your plugin's instance to Handsontable

> ### 2. Extend BasePlugin
> Extending our `BasePlugin` is a recommended way to start creating your next extension.
> In this built-in interface, we care about many factors such as backward compatibility, memory leaks preventions and proper binding plugin's instance with Handsontable.

Reviewer 3: ```suggestion
### 2. Extend the `BasePlugin`
The best way to start creating your own plugin is to extend the `BasePlugin`.

The `BasePlugin` interface takes care of:
* Backward compatibility
* Memory leak prevention
* Properly binding your plugin's instance to Handsontable
```
```

```js
export class CustomPlugin extends BasePlugin {
  /**
   * Define a unique plugin's key.
   * Handsontable uses that string to register the plugin under that alias.
   * Moreover, you might also use the same parameter to recognize the plugin's
   * options passed through the Setting object at Handsontable initialization.
   *
   * @returns {string}

>    */

Reviewer 3: /**
  \* Define a unique key (a string) for your plugin.
   \* The key becomes the plugin's alias.
   \* Handsontable registers the plugin under that alias.
   \* You can also use the alias to recognize the plugin's
   \* options passed through the Setting object
   \* at Handsontable's initialization.
   \*
   \* @returns {string}
   \*/

>   /**
>    * Define a unique plugin's key.
>    * Handsontable uses that string to register the plugin under that alias.
>    * Moreover, you might also use the same parameter to recognize the plugin's
>    * options passed through the Setting object at Handsontable initialization.
>    *
>    * @returns {string}
>    */

Reviewer 3: ```suggestion
 /**
  * Define a unique key (a string) for your plugin.
  * The key becomes the plugin's alias.
  * Handsontable registers the plugin under that alias.
  * You can also use the alias to recognize the plugin's
  * options passed through the Setting object at Handsontable's initialization.
  *
  * @returns {string}
  */
```

  static get PLUGIN_KEY() {
    return 'customPlugin';
  }

  /**
   * Extend the default constructor and define internal properties for your plugin.
   *
   * @param {Handsontable} hotInstance
   */
  constructor(hotInstance) {
    /**
     * BasePlugin's constructor adds to our custom plugin the following properties:
     * - `this.hot` - A reference to the Handsontable instance, you can't overwrite this property.
     *                It also gives you the access to columns and rows index mappers.

>      */

Reviewer 3: \/**
     \* The \`BasePlugin\`'s constructor adds a \`this.hot\` property to your plugin.
     \* The \`this.hot\` property:
     \* - Is a reference to the Handsontable instance.
     \* - Can't be overwritten.
     \* - Gives you access to columns' and rows' index mappers.
\*/

      msg: ''
    };
  }

  /**
   * Unifies configuration passed to settings object.
   *
   * @returns {object}
   * @throws {Error}
   */
  getUnifiedConfig() {
    const pluginSettings = this.hot.getSettings()[CustomPlugin.PLUGIN_KEY];

    if (pluginSettings === true) {
      return {
        enabled: true,
        msg: 'default msg boolean'
      };
    }
    if (Object.prototype.toString.call(pluginSettings) === '[object Object]') {
      return {
        enabled: true,
        msg: 'default msg obj',
        ...pluginSettings
      };
    }
    if (pluginSettings === false) {
      return {
        enabled: false,
        msg: ''
      };
    }

    throw new Error(
      `${CustomPlugin.PLUGIN_KEY} - incorrect plugins configuration.
      Passed:
        - type: ${typeof pluginSettings}
        - value: ${JSON.stringify(pluginSettings, null, ' ')}

      Expected:
        - boolean
        - object
      `
    );
  }

  /**
   * Checks if the plugin is enabled in the settings.
   */
  isEnabled() {
    const pluginSettings = this.getUnifiedConfig();

    return pluginSettings.enabled;
  }

  /**
   * The `enablePlugin` method is triggered on the `beforeInit` hook.

>    * It should contain your initial plugin setup, along with the hooks connections.
>    * Note, that this method is run only if the statement in the `isEnabled` method returns true.

Reviewer 3: ```suggestion
   * It should contain your plugin's initial setup and hook connections.
   * This method is run only if the `isEnabled` method returns `true`.
```

>    */

Reviewer 3: It should contain your plugin's initial setup and hook connections.
This method is run only if the \`isEnabled\` method returns \`true\`.

  enablePlugin() {

>     // Read the plugins' configuration from the initialization object.

Reviewer 3: // Get the plugin's configuration from the initialization object.

    this.configuration = this.getUnifiedConfig();

    // Add all your plugin hooks here. It's a good idea to use arrow functions to keep the plugin as a context.
    this.addHook('afterChange', (changes, source) => this.onAfterChange(changes, source));

>     // The super method assigns the this.enabled property to true.

Reviewer 3: // The \`super\` method sets the \`this.enabled\` property to \`true\`.

    // It is a necessary step to update the plugin's settings properly.
    super.enablePlugin();
  }

  /**

>    * The disablePlugin method is used to disable the plugin.

Reviewer 3: * The \`disablePlugin\` method disables the plugin.

   */
  disablePlugin() {
    // Reset all of your plugin class properties to their default values here.
    this.configuration = null;

    // The `BasePlugin.disablePlugin` method takes care of clearing the hook connections
    // and assigning the 'false' value to the 'this.enabled' property.
    super.disablePlugin();
  }

  /**
   * The updatePlugin method is called on the `afterUpdateSettings` hook
   * (unless the updateSettings method turned the plugin off).
   *
   * It should contain all the stuff your plugin needs to do to work correctly

>    * after updating Handsontable instance settings.

Reviewer 3: \* The \`updatePlugin\` method is called on the \`afterUpdateSettings\` hook
   \* (unless the \`updateSettings\` method turned the plugin off).
   \*
   \* The \`updatePlugin\` method should contain anything that your plugin needs to do to work correctly
   \* after updating the Handsontable instance settings.

   */
  updatePlugin() {
    // The `updatePlugin` method needs to contain all the code needed to properly re-enable the plugin.
    // In most cases simply disabling and enabling the plugin should do the trick.
    const { enabled, msg } = this.getUnifiedConfig();

>     // You can control if settings' update should go through disable->enable routine.

Reviewer 3: // You can decide if updating the settings triggers the the disable->enable routine or not.

    if (enabled === false && this.enabled === true) {
      this.disablePlugin();

    } else if (enabled === true && this.enabled === false) {
      this.enablePlugin();
    }

>     // Because sometimes the only action you need is to update only one option.

Reviewer 3: // If you need to update just a single option.

    if (this.configuration !== null && msg && this.configuration.msg !== msg) {
      this.configuration.msg = msg;
    }

    super.updatePlugin();
  }

  /**
   * Define your external methods.
   */
  externalMethodExample() {
    // Method definition.
  }

  /**
   * The afterChange hook callback.
   *
   * @param {CellChange[]} changes An array of changes.
   * @param {string} source Describes the source of the change.
   */
  onAfterChange(changes, source) {
    // afterChange callback goes here.
    console.log(
      `${CustomPlugin.PLUGIN_KEY}.onAfterChange - ${this.configuration.msg}`,
      changes,
      source
    );
  }

  /**
   * The `destroy` method is the best place to clean up any instances,
   * objects or index mappers created during the plugin's lifecycle.
   */
  destroy() {

>     // The super method takes care of de-assigning the event callbacks,
>     // plugin hooks and clearing all the plugin properties.

Reviewer 3: ```suggestion
    // The `super` method cleans up the plugin's event callbacks, hook connections, and properties.
```

>     super.destroy();

Reviewer 3: // The \`super\` method cleans up the plugin's event callbacks,
// hook connections, and properties.

  }
}
```

### 3. Register CustomPlugin

> Last but not least, you need to register the plugin. This is the only way to use plugins in Handsontable.
> There are two ways to do that.

Reviewer 3: ```suggestion
Now, register your plugin.

There are two ways to register a plugin:
```

> 

Reviewer 3: Now, register your plugin.

There are two ways to register a plugin:

> - You can define a static getter named `PLUGIN_KEY` that the `registerPlugin` utility uses as the plugin's alias. Usage example you can find in the above code snippet.

Reviewer 3: ```suggestion
- **Option 1**: Define a static getter named `PLUGIN_KEY` that the `registerPlugin` utility uses as the plugin's alias. Check the example in the code snippet above.
```

>   ```js

Reviewer 3: \- **Option 1**: Define a static getter named \`PLUGIN_KEY\` that the \`registerPlugin\` utility uses as the plugin's alias. Check the example in the code snippet above.

  // You need to register your plugin in order to use it within Handsontable.
  registerPlugin(CustomPlugin);
  ```

> - You can also use the custom alias. Put a string in the first argument. The registerer will use that string as an alias instead of the `PLUGIN_KEY` getter from `CustomPlugin`.

Reviewer 3: \- **Option 2**: Use a custom alias. Put a string in the first argument. The registerer uses that string as an alias, instead of the \`PLUGIN_KEY\` getter from \`CustomPlugin\`.

> - You can also use the custom alias. Put a string in the first argument. The registerer will use that string as an alias instead of the `PLUGIN_KEY` getter from `CustomPlugin`.

Reviewer 3: ```suggestion
- **Option 2**: Use a custom alias. Put a string in the first argument. The registerer uses that string as an alias, instead of the `PLUGIN_KEY` getter from `CustomPlugin`.
```

  ```js
  registerPlugin('CustomAlias', CustomPlugin);
  ```

> ### 4. Use plugin in Handsontable
> Accordingly, to the supported options for the plugin, you might control it, passing in its options the boolean or object.
> The following snippet presents all possible approaches:

Reviewer 3: ```suggestion
### 4. Use your plugin in Handsontable
To control the plugin's options, pass a boolean or an object at the plugin's initialization:
```

> 

Reviewer 3: \### 4. Use your plugin in Handsontable
To control the plugin's options, pass a boolean or an object at the plugin's initialization:

```js
import Handsontable from 'handsontable';
import { CustomPlugin } from './customPlugin';

const hotInstance = new Handsontable(container, {

>   // You can enable plugin by passing `true` to use default options.

Reviewer 3: // Pass \`true\` to enable the plugin with default options.

  [CustomPlugin.PLUGIN_KEY]: true,
  // You can also enable the plugin by passing an object with options.
  [CustomPlugin.PLUGIN_KEY]: {
    msg: 'user-defined message',
  },
  // You can also initialize the plugin without enabling it at the beginning.
  [CustomPlugin.PLUGIN_KEY]: false,
});
```

### 5. Get a reference to the plugin's instance

> To use the plugin's API you may use [`getPlugin`](../api/core/#getplugin) method to get a reference to its instance.

Reviewer 3: To use the plugin's API, call the [\`getPlugin\`](../api/core/#getplugin) method to get a reference to the plugin's instance.
